File: LocalPayment/src/test/java/com/braintreepayments/api/MockLocalPaymentApiBuilder.java
package com.braintreepayments.api;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;

import org.mockito.stubbing.Answer;

public class MockLocalPaymentApiBuilder {

    private LocalPaymentNonce tokenizeSuccess;
    private Exception tokenizeError;
    private LocalPaymentAuthRequestParams createPaymentMethodSuccess;
    private Exception createPaymentMethodError;

    public MockLocalPaymentApiBuilder tokenizeSuccess(LocalPaymentNonce tokenizeSuccess) {
        this.tokenizeSuccess = tokenizeSuccess;
        return this;
    }

    public MockLocalPaymentApiBuilder tokenizeError(Exception error) {
        this.tokenizeError = error;
        return this;
    }

    public MockLocalPaymentApiBuilder createPaymentMethodSuccess(
            LocalPaymentAuthRequestParams createPaymentMethodSuccess) {
        this.createPaymentMethodSuccess = createPaymentMethodSuccess;
        return this;
    }

    public MockLocalPaymentApiBuilder createPaymentMethodError(Exception error) {
        this.createPaymentMethodError = error;
        return this;
    }

    public LocalPaymentApi build() {
        LocalPaymentApi localPaymentAPI = mock(LocalPaymentApi.class);

        doAnswer((Answer<Void>) invocation -> {
            LocalPaymentInternalTokenizeCallback callback =
                    (LocalPaymentInternalTokenizeCallback) invocation.getArguments()[3];
            if (tokenizeSuccess != null) {
                callback.onResult(tokenizeSuccess, null);
            } else if (tokenizeError != null) {
                callback.onResult(null, tokenizeError);
            }
            return null;
        }).when(localPaymentAPI).tokenize(anyString(), anyString(), anyString(),
                any(LocalPaymentInternalTokenizeCallback.class));

        doAnswer((Answer<Void>) invocation -> {
            LocalPaymentInternalAuthRequestCallback callback =
                    (LocalPaymentInternalAuthRequestCallback) invocation.getArguments()[1];
            if (createPaymentMethodSuccess != null) {
                callback.onResult(createPaymentMethodSuccess, null);
            } else if (createPaymentMethodError != null) {
                callback.onResult(null, createPaymentMethodError);
            }
            return null;
        }).when(localPaymentAPI).createPaymentMethod(any(LocalPaymentRequest.class),
                any(LocalPaymentInternalAuthRequestCallback.class));

        return localPaymentAPI;
    }
}

File: LocalPayment/src/test/java/com/braintreepayments/api/localpayment/MockLocalPaymentApiBuilder.kt
package com.braintreepayments.api.localpayment

import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.mockito.invocation.InvocationOnMock

class MockLocalPaymentApiBuilder {
    private var tokenizeSuccess: LocalPaymentNonce? = null
    private var tokenizeError: Exception? = null
    private var createPaymentMethodSuccess: LocalPaymentAuthRequestParams? = null
    private var createPaymentMethodError: Exception? = null

    fun tokenizeSuccess(tokenizeSuccess: LocalPaymentNonce?): MockLocalPaymentApiBuilder {
        this.tokenizeSuccess = tokenizeSuccess
        return this
    }

    fun tokenizeError(error: Exception?): MockLocalPaymentApiBuilder {
        this.tokenizeError = error
        return this
    }

    fun createPaymentMethodSuccess(
        createPaymentMethodSuccess: LocalPaymentAuthRequestParams?
    ): MockLocalPaymentApiBuilder {
        this.createPaymentMethodSuccess = createPaymentMethodSuccess
        return this
    }

    fun createPaymentMethodError(error: Exception?): MockLocalPaymentApiBuilder {
        this.createPaymentMethodError = error
        return this
    }

    internal fun build(): LocalPaymentApi {
        val localPaymentAPI = Mockito.mock(LocalPaymentApi::class.java)

        Mockito.doAnswer { invocation: InvocationOnMock ->
            val callback =
                invocation.arguments[3] as LocalPaymentInternalTokenizeCallback
            if (tokenizeSuccess != null) {
                callback.onResult(tokenizeSuccess, null)
            } else if (tokenizeError != null) {
                callback.onResult(null, tokenizeError)
            }
            null
        }.`when`(localPaymentAPI).tokenize(
            ArgumentMatchers.anyString(),
            ArgumentMatchers.anyString(),
            ArgumentMatchers.anyString(),
            ArgumentMatchers.any(
                LocalPaymentInternalTokenizeCallback::class.java
            )
        )

        Mockito.doAnswer { invocation: InvocationOnMock ->
            val callback =
                invocation.arguments[1] as LocalPaymentInternalAuthRequestCallback
            if (createPaymentMethodSuccess != null) {
                callback.onLocalPaymentInternalAuthResult(createPaymentMethodSuccess, null)
            } else if (createPaymentMethodError != null) {
                callback.onLocalPaymentInternalAuthResult(null, createPaymentMethodError)
            }
            null
        }.`when`(localPaymentAPI).createPaymentMethod(
            ArgumentMatchers.any(
                LocalPaymentRequest::class.java
            ),
            ArgumentMatchers.any(
                LocalPaymentInternalAuthRequestCallback::class.java
            )
        )

        return localPaymentAPI
    }
}
